packages/eslint-plugin/src/rules/max-params.ts

```diff
@@ -0,0 +1,94 @@
+import { AST_NODE_TYPES, type TSESTree } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+type FunctionLike =
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+type FunctionRuleListener<T extends FunctionLike> = (node: T) => void;
+
+const baseRule = getESLintCoreRule('max-params');
+
+export type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;
+
+export default util.createRule<Options, MessageIds>({
+  name: 'max-params',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Enforce a maximum number of parameters in function definitions',
+      extendsBaseRule: true,
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          maximum: {
+            type: 'integer',
+            minimum: 0,
+          },
+          max: {
+            type: 'integer',
+            minimum: 0,
+          },
+          countVoidThis: {
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [{ max: 3, countVoidThis: false }],
+
+  create(context, [{ countVoidThis }]) {
+    const baseRules = baseRule.create(context);
+
+    if (countVoidThis === true) {
+      return baseRules;
+    }
+
+    const removeVoidThisParam = <T extends FunctionLike>(node: T): T => {
+      if (node.params.length === 0) {
+        return node;
+      }
+
+      const params = [...node.params];
```

Comment:
[Performance] This is probably not impactful for most users, but just in case: it's best not to create arrays unnecessarily. It's a bit extra memory.

Much of this function's _"create a shifted `params` if necessary"_ could be reduced down to:

```ts
const params =
  node.params.length &&
  node.params[0].type === AST_NODE_TYPES.Identifier &&
  node.params[0].name === 'this' &&
  node.params[0].typeAnnotation?.typeAnnotation.type ===
    AST_NODE_TYPES.TSVoidKeyword
    ? node.params.slice(1)
    : node.params;
```

Reply:
yup, I actually had thoughts about performance while writing this, but didn't think it's gonna much of an issue. 
You're totally right (even though it's just a shallow copy of, and let's be real here, _maybe_ 10 items max, but yeah totally agree).

TBH, I don't like huge ternaries, but having it inside a regular condition doesn't look good either, so I just took the cursed ternary path.

Reply:
Yeah to me the important bit is only making a new array if needed.